File: wwwroot/js/container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
const maxAreaBruteForce = function(height) {
    let max = 0;
    for (let i = 0; i < height.length - 1; i++) {
        for (let j = i + 1; j < height.length; j++) {
            let area = (j - i) * Math.min(height[i], height[j]);
            max = Math.max(area, max);
        }
    }

    return max;
};

/**
 * @param {number[]} height
 * @return {number}
 */
 const maxArea = function(height) {
    // For each height, find the furtherest height that is longer than it.
    // Use two pointers.
    if (height=== null || height.length <= 1) {
        return 0;
    }

    let max = 0;
    let i = 0;
    let j = height.length - 1;
    while (i < j) {
        let area = (j - i) * Math.min(height[i], height[j]);
        max = Math.max(max, area);

        if (height[i] > height[j]) {
            j--;
        } else {
            i++;
        }
    }

    return max;
};

document.getElementById("container-with-most-water").innerHTML = maxArea([1,3,2,5,25,24,5]);

File: src/CPP/src/leetcode/ContainerWithMostWater.cpp
/*
 * [Source] https://leetcode.com/problems/container-with-most-water/
 * [Difficulty]: Medium
 * [Tag]: Array
 * [Tag]: Two Pointers
 * [Tag]: Greedy
 */

#include <iostream>

using namespace std;

// [Solution]: for x = 0..n-1, y = n..1, if Ax < Ax-1, skip; if Ay < Ay-1, skip.
// [Corner Case]:
class Solution {
};

// [Solution]: If Ax < Ay, then there is no way to make area become larger if x not change.
/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/arrandstring/ContainerWithMostWater.java
 */

/* Java solution
public class Solution {
    public int maxArea(int[] height) {
        if (height == null || height.length <= 1) {
            return 0; 
        }
        int st = 0; 
        int ed = height.length - 1; 
        int max = Integer.MIN_VALUE; 
        while (st < ed) {
            max = Math.max(max, Math.min(height[st], height[ed]) * (ed - st)); 
            if (height[st] <= height[ed]) {
                ++st; 
            } else {
                --ed; 
            }
        }
        return max; 
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
